libnetwork/iptables/iptables.go

```diff
@@ -173,6 +173,24 @@ func (iptable IPTable) LoopbackByVersion() string {
 	return "127.0.0.0/8"
 }
 
+// SetupInternalICC is used to add or remove the internal interface from the firewalld zone ( ICC )
+// https://github.com/moby/libnetwork/issues/2647
+func (iptable IPTable) SetupInternalICC(bridgeIface string, include bool) error {
+	// Either add or remove the interface from the firewalld zone
+	if firewalldRunning {
+		if include {
+			if err := AddInterfaceFirewalld(bridgeIface); err != nil {
+				return fmt.Errorf("Failed to add interface %s to firewalld zone: %s", bridgeIface, err.Error())
+			}
+		} else {
+			if err := DelInterfaceFirewalld(bridgeIface); err != nil {
+				return fmt.Errorf("Failed to remove interface %s from firewalld zone: %s", bridgeIface, err.Error())
+			}
+		}
+	}
```

Comment:
This code is a near-identical copy of the code at the top of `(IPTable).ProgramChain`. Notably, the only references to `ProgramChain` are in the `(*bridgeNetwork).setupIPTables` function, specifically the `!config.Internal` branch. And the only caller of `setupInternalNetworkRules` is the `config.Internal` branch of `(*bridgeNetwork).setupIPTables`. Given that adding/removing the interface to the firewalld zone is required irrespective of the value of `config.Internal`, I don't think duplicating the same code in both branches is the right solution. I think a less-awful solution would be to make `(*bridgeNetwork).setupIPTables` call `AddInterfaceFirewalld`/`DelInterfaceFirewalld` directly, and move the `if firewalldRunning` checks into `AddInterfaceFirewalld` and `DelInterfaceFirewalld`. There isn't any clean abstraction layer between the bridge driver and package iptables anyway so I see no value in fabricating a half-baked, leaky abstraction which makes no sense unless you know what's behind the abstraction anyway, just to maintain a fiction that there is an abstraction.